app/models/manageiq/providers/hawkular/middleware_manager/alert_profile_manager.rb

```diff
@@ -0,0 +1,110 @@
+      unless old_assignments_ids.empty?
+        to_remove_alerts_ids = old_alerts_ids - new_alerts_ids
+        to_add_alerts_ids = new_alerts_ids - old_alerts_ids
+        to_remove_alerts_ids.each do |alert_id|
+          group_trigger = @alerts_client.get_single_trigger "MiQ-#{alert_id}", true
+          unassign_members(group_trigger, profile_id, old_assignments_ids)
+        end
+        to_add_alerts_ids.each do |alert_id|
+          group_trigger = @alerts_client.get_single_trigger "MiQ-#{alert_id}", true
+          assign_members(group_trigger, profile_id, old_assignments_ids)
+        end
+      end
+    end
+
+    def update_assignments(profile_id, old_alerts_ids, old_assignments_ids, new_assignments_ids)
+      to_unassign_ids = old_assignments_ids - new_assignments_ids
+      to_assign_ids = new_assignments_ids - old_assignments_ids
+      if !to_unassign_ids.empty? || !to_assign_ids.empty?
+        old_alerts_ids.each do |alert_id|
+          group_trigger = @alerts_client.get_single_trigger "MiQ-#{alert_id}", true
+          unassign_members(group_trigger, profile_id, to_unassign_ids) unless to_unassign_ids.empty?
+          assign_members(group_trigger, profile_id, to_assign_ids) unless to_assign_ids.empty?
+        end
+      end
+    end
+
+    def unassign_members(group_trigger, profile_id, members_ids)
+      context, profiles = unassign_members_context(group_trigger, profile_id)
+      group_trigger.context = context
+      @alerts_client.update_group_trigger(group_trigger)
+      if profiles.empty?
+        members_ids.each do |member_id|
+          @alerts_client.orphan_member("#{group_trigger.id}-#{member_id}")
+          @alerts_client.delete_trigger("#{group_trigger.id}-#{member_id}")
+        end
+      end
+    end
+
+    def unassign_members_context(group_trigger, profile_id)
+      context = group_trigger.context.nil? ? {} : group_trigger.context
+      profiles = context['miq.alert_profiles'].nil? ? [] : context['miq.alert_profiles'].split(",")
+      profiles -= [profile_id.to_s]
+      context['miq.alert_profiles'] = profiles.uniq.join(",")
+      [context, profiles]
+    end
+
+    def assign_members(group_trigger, profile_id, members_ids)
+      group_trigger.context = assign_members_context(group_trigger, profile_id)
+      @alerts_client.update_group_trigger(group_trigger)
+      members = @alerts_client.list_members group_trigger.id
+      current_members_ids = members.collect(&:id)
+      members_ids.each do |member_id|
+        next if current_members_ids.include?("#{group_trigger.id}-#{member_id}")
+        create_new_member(group_trigger, member_id)
+      end
+    end
+
+    def assign_members_context(group_trigger, profile_id)
+      context = group_trigger.context.nil? ? {} : group_trigger.context
+      profiles = context['miq.alert_profiles'].nil? ? [] : context['miq.alert_profiles'].split(",")
+      profiles.push(profile_id.to_s) unless profiles.include?(profile_id)
```

Comment:
how big/long is `profiles` typically? `include?` goes over all the array, so it could be inefficient. Using `Set` could be better if profiles is expected to be long.